src/test/java/games/strategy/util/VersionTest.java

```diff
@@ -57,6 +57,20 @@ public void testRead3() {
   }
 
   @Test
+  public void verifyDevVersionToString() {
+    assertEquals("1.2.3.dev", new Version("1.2.3.dev").toString());
+    assertEquals("1_2_3_dev", new Version("1.2.3.dev").toStringFull('_'));
+  }
+
+  @Test
+  public void nullTest() {
+    assertFalse(new Version(1,2,3).isCompatible(null));
+    assertFalse(new Version(1,2,3).equals(null));
+    assertFalse(new Version(1,2,3).isGreaterThan(null));
+    assertFalse(new Version(1,2,3).isLessThan(null));
```

Comment:
Hmm, are the conditions defined by the above two assertions consistent?  The only way a version could be both _not greater than_ and _not less than_ `null` is if it were equal to `null`.  That doesn't seem right when the version under test is obviously non-`null`.

The definition of `compareTo()` made the decision that `null` is always greater than a non-`null` value (not sure I agree with that, but it is what it is), so that for consistency, these tests should be

```java
assertTrue(new Version(1,2,3).isGreaterThan(null));
assertFalse(new Version(1,2,3).isLessThan(null));
```

FWIW, the implementation in `compareTo()` is probably wrong because the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html#compareTo-T-) for `Comparable#compareTo()` explicitly state that `compareTo()` should throw an NPE if its argument is `null` (that's to ensure symmetry between `x.compareTo(null)` and `null.compareTo(x)`).  However, we probably don't want to change that in this PR, as it could have a huge ripple effect on all callers of any method that directly or transitively calls `compareTo()`.

:warning: Actually, it appears the whole implementation of `compareTo()` is the reverse of the interface contract.  For example, `compareTo()` should return a negative value if `this` is less than `other`, but it does the opposite.  The majority of the following tests fail:

```java
@Test
public void testCompareTo() {
  assertThat(new Version(1, 0, 0).compareTo(null), is(lessThan(0)));

  assertThat(new Version(1, 0, 0).compareTo(new Version(2, 0, 0)), is(lessThan(0))); // fails
  assertThat(new Version(1, 0, 0).compareTo(new Version(1, 1, 0)), is(lessThan(0))); // fails
  assertThat(new Version(1, 0, 0).compareTo(new Version(1, 0, 1)), is(lessThan(0))); // fails

  assertThat(new Version(1, 0, 0).compareTo(new Version(1, 0, 0)), is(0));

  assertThat(new Version(2, 0, 0).compareTo(new Version(1, 0, 0)), is(greaterThan(0))); // fails
  assertThat(new Version(1, 1, 0).compareTo(new Version(1, 0, 0)), is(greaterThan(0))); // fails
  assertThat(new Version(1, 0, 1).compareTo(new Version(1, 0, 0)), is(greaterThan(0))); // fails
}
```

As best as I can tell, the only place where `Version#compareTo()` is called outside this class is in `g.s.engine.lobby.client.ui.TableSorter`.  The only consequence to the user is that when they sort on a version column, the data will be sorted in the reverse of what the UI indicates.

Not sure if we should fix this as part of this PR or in a separate PR, but it should be fixed.  Now I understand why the definitions of `isGreaterThan()` and `isLessThan()` appear to not make sense (i.e. they are using the opposite comparison operator than one would expect).  They are apparently hiding the "implementation detail" of our reversed `compareTo()` implementation.

Reply:
My understanding of null comparisons is that they should always return false. That is the case in SQL. e.g. "select 1 from [tbl] where null > 0 or null < 0 or null = 0" will return no rows even when [tbl] has rows.

Do you have a different idea of what nullability should do? None of this should ever happen. An argument could be made that an NPE would give you better debugging information than returning wrong answers. I would have thought that would be preferable in this case.

Reply:
I agree about the test being the opposite of what would normally be expected, but this behaviour is unchanged in this PR. Changing it here would be pretty big scope creep! We'd need to do a bit of work to ensure we don't introduce any problems.

Now, what is holding this one up?

Reply:
> An argument could be made that an NPE would give you better debugging information than returning wrong answers. I would have thought that would be preferable in this case.

I agree.  As I mentioned above, the interface contract for `Comparable#compareTo()` in the Javadocs explicitly states that it should throw an NPE if it receives a `null` argument.

> I agree about the test being the opposite of what would normally be expected, but this behaviour is unchanged in this PR.

Sounds reasonable.  I'll submit a follow-up PR after this is merged to correct the reversed implementation.